src/shimmerHttp.js

```diff
@@ -0,0 +1,215 @@
+}
+
+function wrapHttpGet(mod) {
+  return (options, cb) => {
+    const req = mod.request(options, cb);
+    req.end();
+    return req;
+  };
+}
+
+// these are keys that are node specific and come from the actual js request object
+const unnecessaryReqKeys = [
+  'agent',
+  'automaticFailover',
+  'cache',
+  'decompress',
+  'followRedirect',
+  'retries',
+  'slashes',
+  'strictTtl',
+  'throwHttpErrors',
+  'useElectronNet'
+];
+
+function excludeUnnecessaryReqKeys(obj) {
+  return pickBy(obj, (v, k) => unnecessaryReqKeys.indexOf(k) === -1);
+}
+
+function getReqDataObject(rawOptions) {
+  // options might be a string ie http.get(), coerce to object
+  const data =
+    typeof rawOptions === 'string'
+      ? { href: rawOptions, headers: {} }
+      : Object.assign({}, rawOptions) || {};
+
+  const urlObject =
+    typeof data === 'string' ? url.parse(rawOptions) : rawOptions;
+  Object.assign(data, urlObject);
+
+  // ensure url key is present with full URI
+  data.url = data.href || url.format(data);
+  // simple rename
+  data.query = data.search;
+  // delete duplicate or extraneous keys
+  delete data.search;
+  delete data.host;
+  delete data.href;
+  // delete data['user-agent'];
+
+  // sometimes request headers come in as an array
+  // make them strings to conform to our schema better
+  Object.keys(data.headers || {}).forEach(k => {
+    data.headers[k] = isArray(data.headers[k])
+      ? data.headers[k].join(' ')
+      : data.headers[k];
+  });
+
+  return excludeUnnecessaryReqKeys(data);
+}
+
+const initialResKeys = ['headers', 'statusCode', 'statusMessage'];
+
+function getResDataObject(res) {
+  return pickBy(res, (v, k) => initialResKeys.indexOf(k) > -1);
+}
+
+const defaultKeysToRecord = [
+  'req.headers.user-agent',
+  'req.headers.accept-encoding',
+  'req.method',
+  'req.path',
+  'req.protocol',
+  'req.port',
+  'req.hostname',
+  'req.hash',
+  'req.pathname',
+  'req.url',
+  'req.query',
+  'req.user-agent',
+  'req.accept-encoding',
+  'res.headers.cache-control',
+  'res.headers.content-type',
+  'res.headers.date',
+  'res.headers.etag',
+  'res.headers.strict-transport-security',
+  'res.headers.content-encoding',
+  'res.headers.content-length',
+  'res.headers.age',
+  'res.headers.connection',
+  'res.headers.server',
+  'res.headers.vary',
+  'res.statusCode',
+  'res.statusMessage'
+];
+
+function filterData(config = {}, obj = {}) {
+  if (typeof config.filter === 'function') {
+    return config.filter(obj);
+  }
+  return pickBy(obj, (v, k) => defaultKeysToRecord.indexOf(k) > -1);
+}
+
+function wrapHttpRequest({ timeline, data: moduleData = {}, config = {} }) {
+  return function wrapper(original) {
+    return function execute(rawOptions, originalCallback) {
+      // bail if we have already started tracking this request
+      // this can happen by calling https.request(opts)
+      // which ends up calling http.request(opts)
+      if (originalCallback && originalCallback.__iopipeTraceId) {
```

Comment:
Don't all the `get`, `post`, etc. methods just call `request`? Couldn't we just wrap that?

Reply:
Wrapping the "outer" function is safest/least complex from what I am seeing working with this code. I believe the shimmer library works best with this paradigm.